game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def get_last_move(self):
+        return self._moves_list[-1]
+    
+    def get_connect_number(self):
+        return self._connect_number
+
+    def get_number_of_moves(self):
+        return self._number_of_moves
+    
+    def make_move(self, move, player_next = True):
+               
+        self._board[move[0]][move[1]] = self._player_next
+        self._player_next = - self._player_next
+        self._moves_list.append(move)
+        self._number_of_moves = self._number_of_moves + 1
+
+        # return game_state
+        
+    def unmake_move(self):
+        move = self.get_last_move()
+        self._board[move[0]][move[1]] = None
+        self._player_next = - self._player_next
+        self._number_of_moves = self._number_of_moves - 1
+        self._moves_list = self._moves_list[: -1]
+        
+        # return game_state
+
+    def get_next_moves(self):    
+    
+        global columns
+        global rows
+        global column_order
+                
+        moves = []
+        any_moves = False
+
+        for column in range(columns):
+            if any_moves:
+                break
+            for row in range(rows):
+                if self._board[row][column_order[column]] == None:
+                    moves.append ((row, column_order[column]))
+                    break
+                
+        return moves
+   
+    def game_end(self):
+
+# This determines whether the game has ended after the last move and whether it is a win, lose
+# or draw (if the board is full)
+
+        if self.get_number_of_moves() == 0:
+            return False, False        
+        player = - self.get_player_next()
+        board = self.get_board()
+        connect_number = self.get_connect_number()
+        last_move = self.get_last_move()
+
+        global rows
+        global columns
+        
+        game_end = False
+        win = False

```

Comment:
It's really unclear to me, even reading though this function, how to make sense of the return values here. `win` is presumably a win for player 1? How is a draw represented?

I would probably try and return something else here, maybe an Enum/simple dataclass object, rather than a tuple (even a named-tuple is a bit better than a tuple here imo). For Enum we'd have 4 values of not-over, over-with-draw, p1-win, p2-win